lex.go

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) HashiCorp, Inc.
+
+package mql
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+type lexStateFunc func(*lexer) (lexStateFunc, error)
+
+type lexer struct {
+	source  *bufio.Reader
+	current stack[rune]
+	tokens  chan token
+	state   lexStateFunc
+}
+
+func newLexer(s string) *lexer {
+	l := &lexer{
+		source: bufio.NewReader(strings.NewReader(s)),
+		state:  lexStartState,
+		tokens: make(chan token, 1), // define a ring buffer for emitted tokens
+	}
+	return l
+}
+
+// nextToken is the external api for the lexer and it simply returns the next
+// token or an error. If EOF is encountered while scanning, nextToken will keep
+// returning an eofToken no matter how many times you call nextToken.
+func (l *lexer) nextToken() (token, error) {
+	for {
+		select {
+		case tk := <-l.tokens: // return a token if one has been emitted
+			return tk, nil
+		default: // otherwise, keep scanning via the next state
+			var err error
+			if l.state, err = l.state(l); err != nil {
+				return token{}, err
+			}
+
+		}
+	}
+}
+
+// lexStartState  is the start state.  It doesn't emit tokens, but rather
+// transitions to other states.  Other states typically transition back to
+// lexStartState after they emit a token.
+func lexStartState(l *lexer) (lexStateFunc, error) {
+	panicIfNil(l, "startState", "lexer")
```

Comment:
```suggestion
	panicIfNil(l, "lexStartState", "lexer")
```
To be consistent with the other calls to `panicIfNil`

Reply:
ty!